modules-js/react-fleet/src/form-elements/inputs/MemorableDateInput.tsx

```diff
@@ -0,0 +1,458 @@
+          props.latestDate
+        }`
+      );
+
+      // onlyAllowPast is true, but earliestTime is today or later
+    } else if (props.onlyAllowPast && earliestDateTime >= todayTime) {
+      throw new Error(
+        `When onlyAllowPast is true, earliestDate must be yesterday or earlier.
+          
+          Received: earliestDate: ${props.earliestDate}`
+      );
+
+      // onlyAllowFuture is true, but earliestTime is tomorrow or earlier
+    } else if (props.onlyAllowFuture && earliestDateTime <= todayTime) {
+      throw new Error(
+        `When onlyAllowFuture is true, earliestDate must be tomorrow or later.
+        
+        Received: earliestDate: ${props.earliestDate}`
+      );
+    }
+  }
+
+  if (latestDateTime) {
+    // onlyAllowPast is true, but latestTime is today or later
+    if (props.onlyAllowPast && latestDateTime >= todayTime) {
+      throw new Error(
+        `When onlyAllowPast is true, latestDate must be yesterday or earlier.
+        Received: latestDate: ${props.latestDate}`
+      );
+
+      // onlyAllowFuture is true, but latestTime is today or earlier
+    } else if (props.onlyAllowFuture && latestDateTime <= todayTime) {
+      throw new Error(
+        `When onlyAllowFuture is true, latestDate must be tomorrow or later.
+        Received: latestDate: ${props.latestDate}`
+      );
+    }
+  }
+
+  if (isNotValid && isNotValid.length > 0) {
+    throw new Error(
+      `Initial date is not valid: ${isNotValid}
+    Received: initialDate: ${props.initialDate}`
+    );
+  }
+}
+
+// Determine upper or lower limit date, if applicable.
+export function setDateLimit(
+  suppliedDate: string,
+  hasRelativeLimitation: boolean = false
+): Date | null {
+  const supplied = suppliedDate.length > 0 ? new Date(suppliedDate) : null;
+  const today = new Date(new Date().setHours(0, 0, 0, 0));
+
+  let date: Date | null = null;
+
+  if (supplied) {
+    date = supplied;
+  } else if (hasRelativeLimitation) {
+    date = today;
+  }
+
+  return date;
+}
+
+/**
+ * Returns an error string if user input is incomplete.
+ */
+export function isInputComplete(fields): string {
+  const { month, day, year } = fields;
+  const missingFields: string[] = [];
+
+  let errorString = '';
+
+  if ((month > 0 && month <= 12) === false) {
+    missingFields.push('month');
+  }
+
+  if ((day > 0 && day <= 31) === false) {
+    missingFields.push('day');
+  }
+
+  if ((year.toString().length === 4) === false) {
+    missingFields.push('year');
+  }
+
+  // One field has been skipped
+  if (missingFields.length === 1) {
+    errorString = `Date must include the ${missingFields[0]}`;
+    // Only one field has user input
+  } else if (missingFields.length === 2) {
+    errorString = `Date must include the ${missingFields[0]} and ${
+      missingFields[1]
+    }`;
+    // All fields missing
+  } else if (missingFields.length === 3) {
+    errorString = 'Please enter a date';
+  }
+
+  return errorString;
```

Comment:
This function would be an ideal one to have tests for (explicitly checking that the output says "Date must include the day and year" or whatevs).